File: apps/web/types/pipeline.ts
export type SemVer = string;

export type NodeKind =
  | "dataset"
  | "encoder"
  | "fusion"
  | "objective"
  | "trainer"
  | "evaluator";

export type PortType = string;

export type BlockRef = {
  blockId: string;
  version: SemVer;
};

export type PortDecl = {
  name: string;
  portType: PortType;
  schema?: Record<string, unknown>;
};

export type NodeInstance = {
  id: string;
  type: NodeKind;
  blockRef: BlockRef;
  inputs: PortDecl[];
  outputs: PortDecl[];
  config: Record<string, unknown>;
  ui?: { x: number; y: number };
};

export type EdgeInstance = {
  id: string;
  from: { nodeId: string; port: string };
  to: { nodeId: string; port: string };
};

export type LockedBlock = {
  blockId: string;
  version: SemVer;
  digest: string;
  inputSchema: Record<string, unknown>;
  outputSchema: Record<string, unknown>;
  changelog?: string;
  deprecated?: boolean;
};

export type PipelineSpec = {
  specVersion: "0.1.0";
  pipeline: { id: string; name: string; description?: string; createdAt: string };
  graph: { nodes: NodeInstance[]; edges: EdgeInstance[] };
  lockedBlocks: LockedBlock[];
  runConfig: { seed: number; mode: "sync" | "async" };
};

export type RunCreateResponse = {
  runId: string;
  status: string;
  metrics?: Record<string, unknown>;
};

export type ExplainPortIO = {
  name: string;
  portType: string;
  shape?: number[] | null;
  preview?: number[] | null;
  note?: string | null;
};

export type ExplainStep = {
  nodeId: string;
  nodeType: string;
  blockId: string;
  version: string;
  title: string;
  config: Record<string, unknown>;
  inputs: ExplainPortIO[];
  outputs: ExplainPortIO[];
  formula?: string | null;
  whyItWorks?: string | null;
  impl: string[];
};

export type ExplainResponse = {
  traceVersion: string;
  metrics?: Record<string, unknown>;
  steps: ExplainStep[];
};

File: apps/api/app/schemas/pipeline.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class BlockRef(BaseModel):
    blockId: str = Field(min_length=1)
    version: str = Field(min_length=5)


class PortDecl(BaseModel):
    name: str = Field(min_length=1)
    portType: str = Field(min_length=1)
    schema: Optional[dict[str, Any]] = None


class NodeInstance(BaseModel):
    id: str = Field(min_length=1)
    type: Literal["dataset", "encoder", "fusion", "objective", "trainer", "evaluator"]
    blockRef: BlockRef
    inputs: list[PortDecl] = Field(default_factory=list)
    outputs: list[PortDecl] = Field(default_factory=list)
    config: dict[str, Any] = Field(default_factory=dict)
    ui: Optional[dict[str, Any]] = None


class EdgeEndpoint(BaseModel):
    nodeId: str = Field(min_length=1)
    port: str = Field(min_length=1)


class EdgeInstance(BaseModel):
    id: str = Field(min_length=1)
    from_: EdgeEndpoint = Field(alias="from")
    to: EdgeEndpoint


class Graph(BaseModel):
    nodes: list[NodeInstance]
    edges: list[EdgeInstance] = Field(default_factory=list)


class LockedBlock(BaseModel):
    blockId: str = Field(min_length=1)
    version: str = Field(min_length=5)
    digest: str = Field(min_length=8)
    inputSchema: dict[str, Any] = Field(default_factory=dict)
    outputSchema: dict[str, Any] = Field(default_factory=dict)
    changelog: Optional[str] = None
    deprecated: Optional[bool] = None


class PipelineMeta(BaseModel):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    description: Optional[str] = None
    createdAt: datetime


class RunResources(BaseModel):
    cpuLimit: Optional[float] = None
    memoryMB: Optional[int] = None
    timeoutSec: Optional[int] = None


class RunConfig(BaseModel):
    seed: int = Field(ge=0, default=0)
    mode: Literal["sync", "async"] = "sync"
    resources: Optional[RunResources] = None


class PipelineSpec(BaseModel):
    specVersion: Literal["0.1.0"] = "0.1.0"
    pipeline: PipelineMeta
    graph: Graph
    lockedBlocks: list[LockedBlock]
    runConfig: RunConfig


class RunCreateRequest(BaseModel):
    spec: PipelineSpec


class RunCreateResponse(BaseModel):
    runId: str
    status: str
    metrics: Optional[dict[str, Any]] = None


class RunListItem(BaseModel):
    runId: str
    status: str
    createdAt: datetime
    finishedAt: Optional[datetime] = None
    metrics: dict[str, Any] = Field(default_factory=dict)

